Test the Union Operator on dictionaries

brocla opened the pull request:
Hello @BethanyG .  This PR has four code snippets to test the union operator ( | and |= ) on dictionaries.

a) I got chatGPT to suggest an example :-)   I included my prompt in the file.  It is a function to merge user settings with some default settings. 

b) This one comes from examples in PEP584 (Add Union Operators To dict).

c) The third example comes from the Cpython unit tests.  

d) The last one comes from the `alphametrics` exercise. I modified a call to the .union() method, to use the operator instead.

BethanyG commented:
Hi there @brocla!  Thank you for this.  😄 

I am running behind in life, so haven't gotten to this or your previous PR -- but hopefully will do so before EOD tomorrow. 

-B

BethanyG commented:
Uh. Oh.  This is currently failing CI.  I _think_ that's because the `toml` library has  been deprecated for Python, and Python 3.11+ now has the [tomlib](https://docs.python.org/3.11/library/tomllib.html) module in its place.

![image](https://github.com/exercism/python-representer/assets/5923094/539779fa-7e39-4ab5-b8b6-7688658ab6c2)

I will check later today (_or if you have time, you can_) -- but we probably need to slightly rewrite the ChatGPT example.  Its going off of historical/bad data.  😄

brocla commented:
No problem. Life comes first.

I'll take a look at rewriting with tomlib.

Thanks
Kevin

On Tue, Mar 12, 2024, 5:45 PM BethanyG ***@***.***> wrote:

> Uh. Oh. This is currently failing CI. I *think* that's because the toml
> library has been deprecated for Python, and Python 3.11+ now has the
> tomlib <https://docs.python.org/3.11/library/tomllib.html> module in its
> place.
>
> image.png (view on web)
> <https://github.com/exercism/python-representer/assets/5923094/539779fa-7e39-4ab5-b8b6-7688658ab6c2>
>
> I will check later today (*or if you have time, you can*) -- but we
> probably need to slightly rewrite the ChatGPT example. Its going off of
> historical/bad data. 😄
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/python-representer/pull/53#issuecomment-1992747457>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBULKRFRADZMIWCF5SOEM3YX6HTNAVCNFSM6AAAAABESWKK3CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJSG42DONBVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

brocla commented:
Hmmm....

I converted `toml` to `tomlib` and deleted the leading/trailing blank lines
in the test file.  Now it works.

So that's good...

Kevin

On Tue, Mar 12, 2024 at 7:11 PM Kevin Brown ***@***.***> wrote:

> No problem. Life comes first.
>
> I'll take a look at rewriting with tomlib.
>
> Thanks
> Kevin
>
> On Tue, Mar 12, 2024, 5:45 PM BethanyG ***@***.***> wrote:
>
>> Uh. Oh. This is currently failing CI. I *think* that's because the toml
>> library has been deprecated for Python, and Python 3.11+ now has the
>> tomlib <https://docs.python.org/3.11/library/tomllib.html> module in its
>> place.
>>
>> image.png (view on web)
>> <https://github.com/exercism/python-representer/assets/5923094/539779fa-7e39-4ab5-b8b6-7688658ab6c2>
>>
>> I will check later today (*or if you have time, you can*) -- but we
>> probably need to slightly rewrite the ChatGPT example. Its going off of
>> historical/bad data. 😄
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/exercism/python-representer/pull/53#issuecomment-1992747457>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACBULKRFRADZMIWCF5SOEM3YX6HTNAVCNFSM6AAAAABESWKK3CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJSG42DONBVG4>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

BethanyG commented on test/example-uniondict-normalization/example_uniondict_normalization.py:
```diff
@@ -0,0 +1,94 @@
+
+
+
+""" Examples from PEP584 - Add Union Operators To dict"""
+d = {'spam': 1, 'eggs': 2, 'cheese': 3}
+e = {'cheese': 'cheddar', 'aardvark': 'Ethel'}
+# d | e   # Representer fails, result must be assigned. # TODO Fix this ????
+# e | d   # Representer fails
+x = d | e   # {'spam': 1, 'eggs': 2, 'cheese': 'cheddar', 'aardvark': 'Ethel'}
+y = e | d   # {'cheese': 3, 'aardvark': 'Ethel', 'spam': 1, 'eggs': 2}
+d |= e  # {'spam': 1, 'eggs': 2, 'cheese': 'cheddar', 'aardvark': 'Ethel'}
```
```suggestion
""" Examples adapted from Mecha Munch Management"""

def update_recipes_tuple(ideas, recipe_updates):
    """Update the recipe ideas dictionary.

    :param ideas: dict - The "recipe ideas" dict.
    :param recipe_updates: tuple - tuple with updates for the ideas section.
    :return: dict - updated "recipe ideas" dict.
    """
    
     # Since this action updates the dict in place, 
     # the dict then needs to be returned separately, otherwise it is a syntax error. 
    ideas |= recipe_updates
    return ideas
    
    
    def update_recipes_dcit(ideas, recipe_updates):
    """Update the recipe ideas dictionary.

    :param ideas: dict - The "recipe ideas" dict.
    :param recipe_updates: dict - dictionary with updates for the ideas section.
    :return: dict - updated "recipe ideas" dict.
    """
    
     # Since this action returns a *new* dict, it can go directly on the return line.
    return ideas | recipe_updates
    
##Example Usage##
    
 ideas  = {'Banana Bread' : {'Banana': 1, 'Apple': 1, 'Walnuts': 1, 'Flour': 1, 'Eggs': 2, 'Butter': 1}}
    
 recipe_updates_tuple= (('Banana Bread', {'Banana': 4,  'Walnuts': 2, 'Flour': 1, 'Eggs': 2, 'Butter': 1, 'Milk': 2, 'Eggs': 3}),)
    
recipe_update_dict= {'Banana Bread': {'Banana': 4,  'Walnuts': 2, 'Flour': 1, 'Eggs': 2, 'Butter': 1, 'Milk': 2, 'Eggs': 3}}
    
  update_recipes_tuple(ideas, recipe_updates_tuple)
   {'Banana Bread': {'Banana': 4, 'Walnuts': 2, 'Flour': 1, 'Eggs': 3, 'Butter': 1, 'Milk': 2}}
   
   update_recipes_dict(ideas, recipe_updates_dict)
   {'Banana Bread': {'Banana': 4, 'Walnuts': 2, 'Flour': 1, 'Eggs': 3, 'Butter': 1, 'Milk': 2}}
   
```

brocla replied:
Hi @BethanyG .   I'm been staring at these tests. Perhaps I don't understand what the Golden tests are supposed to accomplish. 

Here is what is bothering me. The `merge_setting()` function and the **Mecha Munch Management** solution use *duck typing* when they use the union operator `|`.   Example:

```
    merged_settings = default_settings | toml_settings
```
`default_settings` and `toml_settings` might be dictionaries, or they might be sets.  The Representer can't know when parsing the code. So, the Representer will be able to parse this,  even if it cannot handle dictionaries with the union operator.

To check if the Representer will parse the new feature, I think we need to not do duck typing. Have code that explicitly uses dictionaries. Something like,
```
    d = {'a': 1} | {'b':2}
```
If the Representer is not handling the new feature, this will fail.

That is where I'm stuck.

Thanks
Kevin

brocla replied:
Oooh. Wait, wait. Better idea.

write the line like this:

```
merged_settings = dict(default_settings) | dict(toml_settings)
```
and also,
```
    ideas |= dict(recipe_updates)
```
That will force the Representer to deal with the dictionaries and the union operator. And we can use the exercise format instead of the little snippets.

What do you think of that?

BethanyG replied:
> _I'v been staring at these tests. Perhaps I don't understand what the Golden tests are supposed to accomplish._

The Golden Tests here are meant to guard against unintended/unexpected _change_ to output.  They're not (exactly) intended to test functionality in the way unit tests or functional/module tests would work, although they can surface bugs that are in those categories as well.

As stated on the [original PR](https://github.com/exercism/python-representer/pull/45) and the PR [it replaced](https://github.com/exercism/python-representer/pull/42), these are a flavor of  _smoke test_.

So **if** the recorded representation is accurate, and the subsequent representations _match_  the originally recorded output, the test passes.  Of course if the original was wrong, and subsequent versions _match_ that wrongness, then the tests also pass.

**Here would be one (_very contrived, mind you_) scenario:**

A contributor decides that for the sake of comparison with other student solutions, we can normalize all instances of `generators` to their short-form `generator-expression` version.  They write an AST transformer to do this, and submit a PR.  That PR fails the CI, because when the [`example-generator-normalization`](https://github.com/exercism/python-representer/blob/main/test/example-generator-normalization/example_generator_normalization.py) runs, its output no longer matches what is on record.  

But it _**doesn't**_ test if the re-write is rewriting correctly (_if the generator and generator expressions are logically the same, or doing the same work_), or if in rewriting we are now matching up solutions that shouldn't be matched up....or if doing the AST transform makes any sense in the first place. 

**Another scenario:**

We upgrade from Python 3.11.8 to Python 3.12.5.  We forget to update [line 86 in `utils.py`](https://github.com/exercism/python-representer/blob/main/representer/utils.py#L86), but remember to add tests for new 3.12 features.  All the new tests fail, but the _existing_ tests pass because the representations are the "same" for everything < 3.12.  In this case, we need to change the line so new features are parsed and represented accurately.

Conversely, we upgrade to 3.12, and **do** update line 86.  The new tests pass, but all of the old ones fail.  So we need to ask: 

1. What changed and why?
2. Is this change a problem, or is it acceptable?
3. Is the change significant enough that we need to re-run all existing student solution representations?

This (obviously) can create problems, and also requires that some (or even all) tests have to be re-recorded when things like OS, Python version, or other tooling changes prompt changes in the code representations.  It is admittedly brittle.  It is also not foolproof or exhaustive when it comes to identifying issues.

Eventually, I would like to have unit and functional tests in place.  But that is further down the road at this point, maybe when the 3.12 change comes, or when I add multi-file processing. The representer has been running in production for ~3+ years, and has not made utter hash of student code -- so it is largely 'OK'.  But we are definitely taking a risk in not having more robust testing, and certainly got caught with our pants down with regards to `astor`.

____

> Here is what is bothering me. The merge_setting() function and the Mecha Munch Management solution use duck typing when they use the union operator |. Example:

    merged_settings = default_settings | toml_settings
> default_settings and toml_settings might be dictionaries, or they might be sets. The Representer can't know when parsing the code. So, the Representer will be able to parse this, even if it cannot handle dictionaries with the union operator.

🤔 


**Very good point.**  And a further thought: there would be no difference between `|` for bitwise operations, `|` for sets, and `|` for dictionaries.  The first needs integers/bits on either side, the second sets, the third dictionaries.  But that only gets checked at runtime, when you call the function and get some flavor of `TypeError` or `SyntaxError`. 

So for this particular instance where we are trying to make sure that we are indeed parsing at a level >= 3.9, we probably do need to be explicit within the function:

```python
def update_recipes_dict_direct():
    """Update the recipe ideas dictionary."""

    # Since this action returns a *new* dict, it can go directly on the return line.
    return ({'Banana Bread': {'Banana': 1, 'Apple': 1, 'Walnuts': 1, 'Flour': 1, 'Eggs': 2, 'Butter': 1}} |
            {'Banana Bread':
                 {'Banana': 4, 'Walnuts': 2, 'Flour': 1, 'Eggs': 2, 'Butter': 1, 'Milk': 2, 'Eggs': 3}})


def update_recipes_tuple_direct():
    """Update the recipe ideas dictionary."""

    ideas = {'Banana Bread': {'Banana': 1, 'Apple': 1, 'Walnuts': 1, 'Flour': 1, 'Eggs': 2, 'Butter': 1}}

    # Since this action updates the dict in place,
    # the dict then needs to be returned separately, otherwise it is a syntax error.
    ideas |= (('Banana Bread', {'Banana': 4, 'Walnuts': 2, 'Flour': 1, 'Eggs': 2, 'Butter': 1, 'Milk': 2, 'Eggs': 3}),)

    return ideas
```

But I've just run into another issue here.  When I went in and changed the target Python version to 3.7, I didn't get the failure I expected.  So I don't know if I need to rebuild the container (probably), or if we have an issue with `ast.parse()`.  So I'm going to play with it some more.

BethanyG replied:
@brocla  -- Apologies, I didn't see your additional comments before I replied.

But I am also having some issues with `ast.parse()`. The docs do [make a point](https://docs.python.org/3/library/ast.html#ast.parse) that there is a "best effort" when using a particular Python version target, and that successful parsing by the AST parser does not guarantee successful compilation.  But I did expect a "barf" or other bad behavior when parsing dictunion/dictupdate operators when targeting Python 3.8, and that is not happening -- even when I rebuild the container with the altered line 86 in `utils.py`.

So this may be a case where the overarching Python version is the determiner of the result.  If that is the case, this is a rather futile test.  Before I say that with confidence, I'd like to test a Python 3.8 container, and also do some manual runs to check the AST output.  Something is definitely up.

brocla replied:
Thanks for the great explanations! This helps me a lot.

I will also try to generate a Representer failure using an older version of python.

BethanyG replied:
Ugh.  Whelp, that was ... fun?  **_Actually, it wasn't._**  
So.  

Did some testing and some digging, and it appears that `ast.parse(feature_version=(<python version>))` is somewhat flakey/problematic/erratic.  You know you're in trouble when you find open issues around changing the docs to be really clear about [**"best effort"**](https://cpython-previews--115980.org.readthedocs.build/en/115980/library/ast.html#ast.parse).

When I tested using `3.8`, `3.9`, `3.10` and `3.11`, it appeared the only true determiner of the AST representation was the main Python version.  Setting `feature_version=(<some version>)` did not appear to have a major effect on AST parsing.  Now, I am pretty tired so I will retry after some sleep - but its looking pretty grim.

The other thing that adds grimness are the following issues I found in the cpython repo:

1.  [cpython issue #115881 comments](https://github.com/python/cpython/issues/115881#issuecomment-1964489116)
2.  [mypy issue #16945](https://github.com/python/mypy/issues/16945)
3. [cpython issue #84794](https://github.com/python/cpython/issues/84794)
4. [cpython issue #115881](https://github.com/python/cpython/issues/115881)
5.  [cpython issue #115980](https://github.com/python/cpython/pull/115980)

Looks like something to do with the rewrite of the Python parser, which happened in Python 3.9 ([PEP Here](https://peps.python.org/pep-0617/)).  There is now a sharp before PEG and after PEG -- and so that nice little flag isn't really set up to switch to an old parser, and so makes a ... guess?  And that guess favors later syntax -- at least for `|` and `|=`.  If you read some of the issues above, it does quite the opposite for other features. 🤦🏽‍♀️ 

That all told, I still think having a test in there is worth it ... but we're probably going to have to let go of truly verifying it, unless we want to set up a Python version matrix.  Again - I am tired, so will need to ponder in the morning.

BethanyG commented on test/example-uniondict-normalization/example_uniondict_normalization.py:
```diff
@@ -0,0 +1,94 @@
+
+
+
+""" Examples from the Cpython unit tests"""
+a = {0: 0, 1: 1, 2: 1}
+b = {1: 1, 2: 2, 3: 3}
+
+c = a.copy()
+c |= b
+
+assert a | b == {0: 0, 1: 1, 2: 2, 3: 3}
+assert c == {0: 0, 1: 1, 2: 2, 3: 3}
+
+c = b.copy()
+c |= a
+
+assert b | a == {1: 1, 2: 1, 3: 3, 0: 0}
+assert c == {1: 1, 2: 1, 3: 3, 0: 0}
+
+c = a.copy()
+c |= [(1, 1), (2, 2), (3, 3)]
+
+assert c == {0: 0, 1: 1, 2: 2, 3: 3}
+
+
+
```
Thinking that if we have two solution-like examples (_if the Mech Munch management suggestion looks good to you_), we can eliminate this one.

BethanyG commented on test/example-uniondict-normalization/example_uniondict_normalization.py:
```diff
@@ -0,0 +1,94 @@
+            else:
+                curr_choices = choices
+
+            for i in curr_choices:
```
```suggestion
            for item in curr_choices:
```

BethanyG commented on test/example-uniondict-normalization/example_uniondict_normalization.py:
```diff
@@ -0,0 +1,94 @@
+                curr_choices = choices
+
+            for i in curr_choices:
+                actual = slc | {letters[0]: i}  # combine two dictionaries
+                new_selections.append((actual, choices - set([i])))
+
```
```suggestion
                actual = slc | {letters[0]: item}  # combine two dictionaries
                new_selections.append((actual, choices - set([item])))

```

BethanyG reviewed (commented):
Nice work!  Added adapted examples from `Mecha Munch Management`, and had a few other suggestions/suggested removals.  LMK if you have questions or issues.  

I can merge later today, after we finalize things.  😄

brocla commented:
Hi Bethany,
  Yes, it might be time to let go. We can use three exercises that we have
already written for unions. They will protect us in the case that an
`astor` version of the Representer sneaks back in.


BTW
  I installed an older version of python (3.8.13) to show that ast would
not parse and unparse a tiny example. But it failed for an unexpected
reason: `ast` has no attribute `unparse`   !?!
It turns out that `unparse()` is new in python 3.9.   I don't think the
python developers are going to spend much time making unparse work
for releases that are older than the feature.

>>> import ast
>>> ast.unparse(ast.parse('{} | {}'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'ast' has no attribute 'unparse'


On Fri, Mar 15, 2024 at 2:41 AM BethanyG ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/example-uniondict-normalization/example_uniondict_normalization.py
> <https://github.com/exercism/python-representer/pull/53#discussion_r1525923777>
> :
>
> > +""" Examples from PEP584 - Add Union Operators To dict"""
> +d = {'spam': 1, 'eggs': 2, 'cheese': 3}
> +e = {'cheese': 'cheddar', 'aardvark': 'Ethel'}
> +# d | e   # Representer fails, result must be assigned. # TODO Fix this ????
> +# e | d   # Representer fails
> +x = d | e   # {'spam': 1, 'eggs': 2, 'cheese': 'cheddar', 'aardvark': 'Ethel'}
> +y = e | d   # {'cheese': 3, 'aardvark': 'Ethel', 'spam': 1, 'eggs': 2}
> +d |= e  # {'spam': 1, 'eggs': 2, 'cheese': 'cheddar', 'aardvark': 'Ethel'}
>
> Ugh. Whelp, that was ... fun? *Actually, it wasn't.*
> So.
>
> Did some testing and some digging, and it appears that ast.parse(feature_version=(<python
> version>)) is somewhat flakey/problematic/erratic. You know you're in
> trouble when you find open issues around changing the docs to be really
> clear about *"best effort"*
> <https://cpython-previews--115980.org.readthedocs.build/en/115980/library/ast.html#ast.parse>
> .
>
> When I tested using 3.8, 3.9, 3.10 and 3.11, it appeared the only true
> determiner of the AST representation was the main Python version. Setting feature_version=(<some
> version>) did not appear to have a major effect on AST parsing. Now, I am
> pretty tired so I will retry after some sleep - but its looking pretty grim.
>
> The other thing that adds grimness are the following issues I found in the
> cpython repo:
>
>    1. cpython issue #115881 comments
>    <https://github.com/python/cpython/issues/115881#issuecomment-1964489116>
>    2. mypy issue #16945 <https://github.com/python/mypy/issues/16945>
>    3. cpython issue #84794
>    <https://github.com/python/cpython/issues/84794>
>    4. cpython issue #115881
>    <https://github.com/python/cpython/issues/115881>
>    5. cpython issue #115980
>    <https://github.com/python/cpython/pull/115980>
>
> Looks like something to do with the rewrite of the Python parser, which
> happened in Python 3.9 (PEP Here <https://peps.python.org/pep-0617/>).
> There is now a sharp before PEG and after PEG -- and so that nice little
> flag isn't really set up to switch to an old parser, and so makes a ...
> guess? And that guess favors later syntax -- at least for | and |=. If
> you read some of the issues above, it does quite the opposite for other
> features. 🤦🏽‍♀️
>
> That all told, I still think having a test in there is worth it ... but
> we're probably going to have to let go of truly verifying it, unless we
> want to set up a Python version matrix. Again - I am tired, so will need to
> ponder in the morning.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/python-representer/pull/53#discussion_r1525923777>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBULKQZ34QNT6TGTOPTS23YYKXZLAVCNFSM6AAAAABESWKK3CVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMZYGM3TKNZTGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

BethanyG commented on test/example-uniondict-normalization/example_uniondict_normalization.py:
```diff
@@ -0,0 +1,97 @@
+    
+     # Since this action updates the dict in place, 
+     # the dict then needs to be returned separately, otherwise it is a syntax error. 
+    ideas |= dict(recipe_updates)
```
```suggestion
    ideas |= recipe_updates
```

BethanyG commented on test/example-uniondict-normalization/example_uniondict_normalization.py:
```diff
@@ -0,0 +1,97 @@
+    :return: dict - updated "recipe ideas" dict.
+    """
+    
+     # Since this action updates the dict in place, 
```
```suggestion
     # recipe_updates here is a tuple.
     # Since this action updates the dict in place, 
```

BethanyG commented:
> Hi Bethany, Yes, it might be time to let go. We can use three exercises that we have already written for unions. They will protect us in the case that an `astor` version of the Representer sneaks back in. BTW I installed an older version of python (3.8.13) to show that ast would not parse and unparse a tiny example. But it failed for an unexpected reason: `ast` has no attribute `unparse` !?! It turns out that `unparse()` is new in python 3.9. I don't think the python developers are going to spend much time making unparse work for releases that are older than the feature.
> >> import ast >> ast.unparse(ast.parse('{} | {}'))
> Traceback (most recent call last): File "<stdin>", line 1, in <module> AttributeError: module 'ast' has no attribute 'unparse'
> […](#)


Yup.  💙   I also ran into that.  And since 3.7 is already EOL, and 3.8 is due to EOL in October, well....

So I agree:  we have some good examples, and we've both beaten and tested the "dead horse", so to speak.  😆   I am going to merge as soon as I fix the CI.

brocla commented:
Yeah! Great news. Thank you.

Kevin


... I'll go start another one.

On Fri, Mar 15, 2024, 4:09 PM BethanyG ***@***.***> wrote:

> Merged #53 <https://github.com/exercism/python-representer/pull/53> into
> main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/python-representer/pull/53#event-12137196146>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBULKQPTZINU5RBBVGVA2DYYNWQVAVCNFSM6AAAAABESWKK3CVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJSGEZTOMJZGYYTINQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
